internal/controller/dnsrecord_healthchecks.go

```diff
@@ -2,180 +2,161 @@ package controller
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%s", dnsRecord.Name, leaf),
+				Namespace: dnsRecord.Namespace,
+				Labels:    map[string]string{ProbeOwnerLabel: BuildOwnerLabelValue(dnsRecord)},
+			},
+			Spec: v1alpha1.DNSHealthCheckProbeSpec{
+				Port:                     dnsRecord.Spec.HealthCheck.Port,
+				Hostname:                 dnsRecord.Spec.RootHost,
+				Address:                  leaf,
+				Path:                     dnsRecord.Spec.HealthCheck.Path,
+				Protocol:                 dnsRecord.Spec.HealthCheck.Protocol,
+				Interval:                 dnsRecord.Spec.HealthCheck.Interval,
+				AdditionalHeadersRef:     dnsRecord.Spec.HealthCheck.AdditionalHeadersRef,
+				FailureThreshold:         dnsRecord.Spec.HealthCheck.FailureThreshold,
+				AllowInsecureCertificate: allowInsecureCerts,
+			},
+		})
 	}
+	return probes
 }
 
-// idForEndpoint returns a unique identifier for an endpoint
-func idForEndpoint(dnsRecord *v1alpha1.DNSRecord, endpoint *externaldns.Endpoint, address string) (string, error) {
-	hash := md5.New()
-	if _, err := io.WriteString(hash, fmt.Sprintf("%s/%s@%s:%s-%v", dnsRecord.Name, endpoint.SetIdentifier, endpoint.DNSName, address, dnsRecord.Generation)); err != nil {
-		return "", fmt.Errorf("unexpected error creating ID for endpoint %s", endpoint.SetIdentifier)
+// BuildOwnerLabelValue ensures label value does not exceed the 63 char limit
+// It adds namespace and name of the record,
+// if the resulting string longer than 63 chars it attempts the following in
+// a stated order and uses the first solution that produces value of less than 63 characters:
+//
+// 1. Trims namespace part to get under the limit.
+// Result: "short-namespace_hostname"
+//
+// 2. Uses the first two subdomains of host instead of the name (e.g. will use "pat.the" from "pat.the.cat.com").
+// Result: "original-namespace_pat.the"
+//
+// 3. Uses GetUIDHash() of v1alpha1.DNSRecord struct
+// Result: "UIDHash"
+func BuildOwnerLabelValue(record *v1alpha1.DNSRecord) string {
+	value := fmt.Sprintf("%s_%s", record.Namespace, record.Name)
+
+	// using the name of the dns record (hostname) and namespace is likely to exceed a 63 char limit
+	if len(value) > 63 {
+		// determine how much we need to remove
+		overshoot := len(value) - 63
+
+		// if we can fix it by trimming NS
+		if len(record.Namespace) > overshoot {
+			value = fmt.Sprintf("%s_%s", record.Namespace[:len(record.Namespace)-overshoot], record.Name)
+		} else {
+			// trimming namespace is not an option - too long hostname
+			// the name of the probe will be fine - it has a limit of 253
+			shortHost := strings.Join(strings.Split(record.Name, ".")[:3], ".")
+
+			// if this the case we can reduce just host
+			if len(record.Name)-len(shortHost) > overshoot {
+				value = fmt.Sprintf("%s_%s", record.Namespace, shortHost)
+			} else {
+				// we can't deal with it just by shortening one of them
+				// default to UID of record
+				value = record.GetUIDHash()
```

Comment:
why don't we just use this all the time?

Reply:
> actually the pobe will always be in the same namespace as the record. The record name has to be unique per namespace. So the label can just be the name of the DNSRecord it came from right? Don't need namespace here ?

if we have multiple namespaces and the same names of records among them this might case a problem for the controller. Right now I'm [restricting](https://github.com/Kuadrant/dns-operator/pull/271/files#diff-d5275729eb4df664c962c5be81ebe5010e601471d12429fcc42376d6d440a142R59) it to a namespace but I'd rather have it safe and have those label values be as unique as we can

Reply:
however, if that makes no sense I'd rather have this logic as simple as possible (either name of record or uid if the name is too long). I might be just too paranoid :thinking:

Reply:
I think just go with the hash as the simplest option for the controller to use and add an additional label of the name for the human? When you do the list you could also restrict to namespace. So the list of probes is always done in the same namespace as the policy and record

Reply:
changed